src/runtime/backend_loader.cpp

```diff
@@ -0,0 +1,178 @@
+    char *err = dlerror();
+    if (err) {
+      HIPSYCL_DEBUG_ERROR << err << std::endl;
+    }
+
+    return false;
+  } else {
+    void *symbol = dlsym(handle, "hipsycl_backend_plugin_get_name");
+
+    char *err = dlerror();
+    if (err) {
+      HIPSYCL_DEBUG_ERROR << "backend_loader: Could not retrieve backend name"
+                          << err << std::endl;
+    }
+    else {
+      auto get_name =
+          reinterpret_cast<decltype(&hipsycl_backend_plugin_get_name)>(symbol);
+
+      handle_out = handle;
+      backend_name_out = get_name();
+
+      return true;
+    }
+  }
+
+  // In case of failure, close handle to avoid resource leak
+  close_plugin(handle);
+  return false;
+}
+
+hipsycl::rt::backend *create_backend(void *plugin_handle) {
+  assert(plugin_handle);
+
+  void *symbol = dlsym(plugin_handle, "hipsycl_backend_plugin_create");
+  char *err = dlerror();
+  if (err) {
+    HIPSYCL_DEBUG_ERROR
+        << "backend_loader: Could not find symbol for backend creation" << err
+        << std::endl;
+    
+    return nullptr;
+  }
+  
+  auto create_backend_func =
+      reinterpret_cast<decltype(&hipsycl_backend_plugin_create)>(symbol);
+
+  return create_backend_func();
+}
+
+}
+
+namespace hipsycl {
+namespace rt {
+
+void backend_loader::query_backends() {
+  std::string install_prefix = HIPSYCL_INSTALL_PREFIX;
+
+  std::filesystem::path backend_lib_path =
+      std::filesystem::path{install_prefix} / "lib/hipSYCL";
+
+  std::string shared_lib_extension = ".so";
+  
+  for (const std::filesystem::directory_entry &entry :
+       std::filesystem::directory_iterator(backend_lib_path)) {
+
+    if (entry.is_regular_file()) {
+      auto p = entry.path();
+      if (p.extension().string() == shared_lib_extension) {
+        std::string backend_name;
+        void *handle;
+        if (load_plugin(p.string(), handle, backend_name)) {
+          HIPSYCL_DEBUG_INFO << "Successfully opened plugin: " << p
+                             << " for backend '" << backend_name << "'"
+                             << std::endl;
+          _handles.emplace_back(std::make_pair(backend_name, handle));
+        }
+      }
+    }
+    
+  }
+}
+
+backend_loader::~backend_loader() {
+  for (auto &handle : _handles) {
+    assert(handle.second);
+    close_plugin(handle.second);
+  }
+}
+
+std::size_t backend_loader::get_num_backends() const { return _handles.size(); }
+
+std::string backend_loader::get_backend_name(std::size_t index) const {
+  assert(index < _handles.size());
+  return _handles[index].first;
+}
+
+bool backend_loader::has_backend(const std::string &name) const {
+  for (const auto &h : _handles) {
+    if (h.first == name)
+      return true;
+  }
+
+  return false;
+}
+
+backend *backend_loader::create(std::size_t index) {
+  assert(index < _handles.size());
+  
+  return create_backend(_handles[index].second);
+}
+
+backend *backend_loader::create(const std::string &name) {
+  if (!has_backend(name))
```

Comment:
this is checked in the for loop as well.. no reason to go through the backends twice

Reply:
Done :)